packages/SwingSet/test/upgrade/test-upgrade.js

```diff
@@ -522,7 +531,7 @@ async function testMultiKindUpgradeChecks(t, mode, complaint) {
   const [v2status, v2result] = await run('upgradeV2Simple', [mode]);
   if (complaint) {
     t.is(v2status, 'rejected');
-    const e = kunser(kunser(v2result));
+    const e = kunser(v2result);
```

Comment:
Did you actually run this updated test?  I'm pretty sure the double deserialization is necessary here because the error is itself a serialized payload.

Reply:
Yep, and both CI and local tests pass. The matching change is in kernel.js line 835 to avoid double-*serializing* the error. This:
```js
    function makeFailure(_errorCD) {
      insistCapData(_errorCD); // kser(Error)
      // const error = kunser(_errorCD)
      // actually we shouldn't reveal the details, so instead we do:
      const error = Error('vat-upgrade failure');
      return ['vatUpgradeCallback', [upgradeID, false, error]];
    }
```

used to be:
```js
      const error = kser(Error('vat-upgrade failure'));
```

but if you trace through how the return value of `makeFailure` is used, it gets stashed in `vatAdminMethargs`, which eventually gets passed into `queueToKref()`, which (thanks to your excellent change) now performs it's own serialization (i.e. `queueToKref` now takes ersatz Presences, not capdata):

(kernel.js line 1267)
```js
    if (crankResults.vatAdminMethargs) {
      // we use terminateVat() to notify vat-admin about failed vat
      // creation, but vatAdminMethargs for successful vat creation,
      // and failed/successful vat upgrades.
      const [method, args] = crankResults.vatAdminMethargs;
      queueToKref(vatAdminRootKref, method, args, 'logFailure');
    }
```

The double-serialization was a bug: when userspace received the rejection value of `upgradeVat()`, it would get capdata of an Error, when it should have gotten an Error.

I wish typescript would help here, but the `args` input to `queueToKref()` is now quite liberal (whereas previously it wanted capdata), which is more ergonomic to use, but harder to check a type on.

The test cases (the ones I modified to change `kunser(kunser(results))` into just `kunser(results)`) don't make that clear. They tell userspace (`tests/upgrade/bootstrap-upgrade.js`) to run `upgradeV2WhichExplodes()` and return an array of events:

```js
    upgradeV2WhichExplodes: async () => {
      const bcap = await E(vatAdmin).getNamedBundleCap('ulrik2');
      const events = [];
      const handler = Far('handler', { ping: msg => events.push(msg) });
      const explode = 'kaboom';
      const vatParameters = { youAre: 'v2', marker, handler, explode };
      // vp.handler causes v2 to handler~.ping(), but that gets unwound
      const p = E(ulrikAdmin).upgrade(bcap, { vatParameters }); // throws
      await p.catch(e => events.push(e));
      await E(ulrikRoot).pingback(handler); // goes to post-rewind v1
      return events;
    },
```

and then `test-upgrade.js` inspects the (serialized) `events` that it returns, one of which is expected to be the Error. `test-upgrade.js` is not inside the vat, and the `c.kpResolution()` it uses gives back capdata, so it must do a `kunser()` to get something equivalent to what the userspace/vat would get. That kinda concealed the fact that a double `kunser` is wrong.

This would be more obvious if the userspace code (in `bootstrap-upgrade.js`) checked `e` for being an `Error` directly (e.g. if it does a check and then includes a boolean in the `events` array with that claim). That would give us an assertion that doesn't have a `kunser` in the pathway. To make *that* test pass with the incorrect double-serialization, you would have needed to add a `kunser()` into the *vat* code, and *that* would have been obviously wrong.

I'll push another commit to change the test that way, which should help make things less ambigous.